vm/src/stdlib/builtins.rs

```diff
@@ -248,11 +248,37 @@ mod builtins {
     #[cfg(feature = "rustpython-compiler")]
     #[pyfunction]
     fn eval(
-        source: Either<PyStrRef, PyRef<crate::builtins::PyCode>>,
+        source: Either<ArgStrOrBytesLike, PyRef<crate::builtins::PyCode>>,
         scope: ScopeArgs,
         vm: &VirtualMachine,
     ) -> PyResult {
-        run_code(vm, source, scope, compile::Mode::Eval, "eval")
+        // source as string
+        let code = match source {
+            Either::A(either) => {
+                let source: &[u8] = &either.borrow_bytes();
+                for x in source {
+                    if *x == 0 {
+                        return Err(vm.new_value_error(
+                            "source code string cannot contain null bytes".to_owned(),
+                        ));
+                    }
+                }
+
+                match std::str::from_utf8(source) {
+                    Ok(s) => Ok(Either::A(vm.ctx.new_str(s))),
+                    Err(err) => {
+                        let msg = format!(
+                            "(unicode error) 'utf-8' codec can't decode byte 0x{:x?} in position {}: invalid start byte",
+                            source[err.valid_up_to()],
+                            err.valid_up_to()
+                        );
+                        Err(vm.new_exception_msg(vm.ctx.exceptions.syntax_error.to_owned(), msg))
+                    }
```

Comment:
I prefer to handle error first not to disturb to read main control flow
```suggestion
                let source = std::str::from_utf8(source).map_err(|err| {
                    let msg = format!(
                            "(unicode error) 'utf-8' codec can't decode byte 0x{:x?} in position {}: invalid start byte",
                            source[err.valid_up_to()],
                            err.valid_up_to()
                        );
                        Err(vm.new_exception_msg(vm.ctx.exceptions.syntax_error.to_owned(), msg))
                })?;
                Either::A(vm.ctx.new_str(s))),

```

Reply:
I applied it in [c672d8f](https://github.com/RustPython/RustPython/pull/4052/commits/c672d8f8b2217d41b680eeeb7dbe38b4f9a1119c).